ariadne/asgi/graphql.py

```diff
@@ -0,0 +1,129 @@
+from typing import Optional, Dict, Type
+from datetime import timedelta
+
+from graphql import GraphQLSchema
+from starlette.types import Receive, Scope, Send
+from starlette.websockets import WebSocket
+
+from ..format_error import format_error
+from .handlers import GraphQLTransportWS, GraphQLWS, GraphQLHTTP, GraphQLWebsocketBase
+from ..types import (
+    ContextValue,
+    ErrorFormatter,
+    RootValue,
+    ValidationRules,
+    Extensions,
+    Middlewares,
+    OnComplete,
+    OnConnect,
+    OnDisconnect,
+    OnOperation,
+    SubscriptionHandlers,
+)
+
+
+class GraphQL:
+    def __init__(
+        self,
+        schema: GraphQLSchema,
+        *,
+        context_value: Optional[ContextValue] = None,
+        root_value: Optional[RootValue] = None,
+        on_connect: Optional[OnConnect] = None,
+        on_disconnect: Optional[OnDisconnect] = None,
+        on_operation: Optional[OnOperation] = None,
+        on_complete: Optional[OnComplete] = None,
+        validation_rules: Optional[ValidationRules] = None,
+        debug: bool = False,
+        introspection: bool = True,
+        logger: Optional[str] = None,
+        error_formatter: ErrorFormatter = format_error,
+        extensions: Optional[Extensions] = None,
+        middleware: Optional[Middlewares] = None,
+        keepalive: float = None,
+        connection_init_wait_timeout: timedelta = timedelta(minutes=1),
+        subscription_handlers: Optional[
+            SubscriptionHandlers[GraphQLWebsocketBase]
+        ] = None,
+    ):
+        self.context_value = context_value
+        self.root_value = root_value
+        self.on_connect = on_connect
+        self.on_disconnect = on_disconnect
+        self.on_operation = on_operation
+        self.on_complete = on_complete
+        self.validation_rules = validation_rules
+        self.debug = debug
+        self.introspection = introspection
+        self.logger = logger
+        self.error_formatter = error_formatter
+        self.extensions = extensions
+        self.middleware = middleware
+        self.keepalive = keepalive
+        self.schema = schema
+        self.connection_init_wait_timeout = connection_init_wait_timeout
+        if not subscription_handlers:
+            self.subscription_handlers: Dict[str, Type[GraphQLWebsocketBase]] = {
+                GraphQLTransportWS.PROTOCOL: GraphQLTransportWS,
+                GraphQLWS.PROTOCOL: GraphQLWS,
+            }
+        else:
+            self.subscription_handlers = {
+                handler.PROTOCOL: handler for handler in subscription_handlers
+            }
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send):
+        if scope["type"] == "http":
+            await GraphQLHTTP(
```

Comment:
Please factor this branch into separate method.

We need to be able to provide custom `GraphQLHTTP` implementation through option.

Reply:
The way it is done now makes the code backwards compatible but less elastic indeed. I can refactor this to something like below:

`http_handler = GraphQLHttpHandler(schema, error_formatter=error_formatter)`

`subscription_handler1 = GraphQLWSHandler(schema, error_formatter=error_formatter, http_handler=http_handler)`
`subscription_handler2 = GraphQLTransportWSHandler(schema, error_formatter=error_formatter, http_handler=http_handler)`

`app1 = GraphQL(http_handler=http_handler, subscription_handler=subscription_handler1)`
`app2 = GraphQL(http_handler=http_handler, subscription_handler=subscription_handler2)`

The code is much simpler but it is a breaking change. What do you think?

Extra note: passing **http_handler** to, eg. **GraphQLWSHandler** as a parameter will make it possible to handle queries/mutations over websocket.

Reply:
I've refactored the code according to the proposal above. I've not added the default HTTP handler though. This would require passing all the parameters needed by the GraphQLHttpHandler to the GraphQL constructor.

With explicit handlers:
  
```
http_handler = GraphQLHttpHandler(schema, error_formatter=error_formatter)
app1 = GraphQL(http_handler=http_handler, subscription_handler=subscription_handler1)
```

with default HTTP handler created by GraphQL class (it is not implemented this way) this would be:

`app1 = GraphQL(schema=schema, error_formatter=error_formatter, subscription_handler=subscription_handler1)`

We would possibly need something like the below to implement the above:

```
def __init__(self, *args, **kwargs):
     self.http_handler = kwargs['http_handler'] if 'http_handler' in kwargs else GraphQLHttpHandler(*args, **kwargs)
     self.subscription_handler = kwargs['subscription_handler'] if 'subscription_handler' in kwargs else GraphQLWsHandler(*args, **kwargs)
```

It seems to me that a more explicit way of passing handlers and parameters is better so I've not implemented the above.